Proposal for an 'allowed widgets' setting

turt2live opened the pull request.

turt2live committed:
Proposal for an 'allowed widgets' setting

manuroe commented:
Will it be possible to trust widgets per domain?
Needing to trust every single widget is a bit boring, no? Instead a user could trust all widgets coming from the some vendor or host.

turt2live commented:
Trusting per domain is unreliable because the widget itself is likely to be wrapped by an integration manager. If you were to trust all of scalar.vector.im for instance, it would potentially open you up to an attacker's custom widget which steals your data. 

Partly because of this reason, riot-web has been requiring each widget to be individually accepted. We do have the additional feature where if the logged in user is also the creator of the widget then we don't require them to approve the widget (they created it, so they implicitly trust it) - this is more of an implementation detail for the user story though.

turt2live committed:
Coalesce widget IDs into a single event

Due to complications in trying to read the events, it's a bit better to just store all the widget IDs in one place.

turt2live commented:
@manuroe I've changed this to put all the widget IDs into one event rather than spreading them out. Please let me know what you think.

This is mostly because trying to implement the `im.vector.setting.widget.<widget ID>` event in riot-web is difficult, and this new structure should be a bit simpler.

manuroe commented:
> Trusting per domain is unreliable

Sure. I guess we are not at the point we can trust an integration manager provider for that. But in the future, such provider could curate, validate and sign widgets they propose to the user.

Note that nothing prevents malicious update of a widget already accepted by the the user.


As widgets instances have unique ids, if a user accepts an instance of a room widget called "The super widget" in a room, they will have to accept again and again any other instance of "The super widget" they will meet after.
Is it something we want?
We could trust urls to fix this.

I am still trying to check how to make use of widgets for users as smooth as possible because I barely see users accepting every single widget.

manuroe commented:
> Is it something we want?

The answer is yes. We made this decision so that we do not leak the user id within the the widget url.

turt2live committed:
Incorporate the widget's URL into the key and add an example

turt2live commented:
and promptly unblocked

lampholder commented on spec/settings.md:
````diff
@@ -117,22 +117,61 @@ settings. Other clients should respect this and not show the prompt if the user
+would result in the following room account data `content`:
+```json
+{
+    "customwidget_%40alice%3Aexample.org_1543007630924_https://scalar.vector.im/api/widgets/generic.html?url=$curl&room_id=$matrix_room_id": true
````
Did you consider something like:

```
{
"customwidget_%40alice%3Aexample.org_1543007630924": "c420e6fe0b017a22768f0a44f4beddcf8e6e2793"
}
```

Where c420e6fe0b017a22768f0a44f4beddcf8e6e2793 is shasum({the widget object}) ?

This would allow us at some point in future say that the widget needs new permission because it's changed (rather than just presenting it as a whole new widget).

turt2live replied:
I'd rather not have all the Riot clients agree on a canonical json implementation in order to achieve that stably. 

The substantial part of the widget is the URL - the rest is purely aesthetic.

lampholder replied:
A canonical json implementation of the setting? Or the widget?

_Is_ the URL the only material part? If we're not including the event ID or the room ID what stops somebody transporting an approved widget to a room in which the approval is dangerous? Is the 'data' not also material?

turt2live replied:
To get consistent hashes we'd need to format the object the same way. Canonical JSON is how we do this elsewhere.

The room ID is implicit by nature of this being a room account data setting.

Using the event ID is probably fine, but still annoying to look up. It'll also result in more warnings as people only change titles. We also don't tell people what `data` contains because it's hard-coded values that usually end up in the URL. The personal information variables are not included in `data`.

lampholder replied:
Ah - do we not just serve the JSON as it is written in the database? I thought it wouldn't matter if the clients wrote the JSON differently as long as it could be read "as it was written", but if that's not possible it's probably not worth the faff.

Personal information isn't included in data, but if the data represents the configuraiton of the widget and if the configuration of the widget could result in personal data being shared/accessible in a different scope then I think it's worth including in the representation of the widget's identity.

lampholder replied:
> The room ID is implicit by nature of this being a room account data setting.

Understood

turt2live replied:
> Ah - do we not just serve the JSON as it is written in the database?

JSON keys have no specific ordering rules. Even with Java/Kotlin the keys will end up in a different order from JavaScript. This is where canonical JSON fixes the problem.

> ...if the configuration of the widget could result in personal data being shared/accessible in a different scope...

It can't. The personal information (avatar, room ID, etc) is passed as spec-defined variables. The other variables are for the purposes of clients which natively render the widgets (which by the way might be another question) - the widget itself is kept the same.

lampholder replied:
> JSON keys have no specific ordering rules.

Understood.

> It can't.

I was looking at the example and the data section looked like: 
```
    "data": {
        "curl": "https://matrix.org"
    },
```
Which made me think that even if the data you're sending to the widget isn't changing, if via the `data` section you could materially alter the widget's behaviour, your data _could_ end up being shared in a different way.

If the `data` section doesn't work that way then fine :)

turt2live replied:
fuuuuuuuuuuuu-

yes, that is how it works and I've been wrong this whole time. I forgot that we put URLs and crap in there :/

I guess we just track the event ID and accept that the user will be extra prompted sometimes?

lampholder replied:
> I guess we just track the event ID and accept that the user will be extra prompted sometimes?

Yeah, sometimes it'll be a cosmetic change that triggers a prompt; I think that's fine for now (better to fail safe, at least).

turt2live committed:
Track by event ID instead

One day we'll figure this out

manuroe commented:
Do you think we can add some granularity in `im.vector.setting.allowed_widgets` to make usages we talked about easier to implement in the future?

We could add an additional layer like `widgets` in `im.vector.setting.allowed_widgets` where we store the data specified in this current spec. Then, we will be able to add more sources of trust like rooms, vendors, senders, etc which will be specified when needed.

That will look like:
```json
{
    "widgets": { 
        "$1543854381213sKqbg:example.org": true
    },
    "vendors": {
        "https://modular.im/"
    }
}
```

turt2live committed:
Allow for future features

turt2live commented:
@manuroe added, though I'm seriously unconvinced that we'll need the `vendors` support. It's a security risk waiting to happen :(

dbkr commented on spec/settings.md:
````diff
@@ -109,3 +109,70 @@ should be set to `false`. This should have no bearing on whether or not widgets
+}
+```
+
+Yes, the key is very verbose.
````
It's not so verbose anymore

dbkr reviewed (commented):
How is the user going to be prompted to accept a widget? What information will be presented to them? This implies they need to see and approve the contents of the 'data' object too, and therefore know how any given widget will behave (ie. that 'custom widget' will load the page in `curl`). How is the user going to make a decision on this?

I can't help but think that if the user is going to have to re-approve the widget every time it's edited, this is only incrementally better than them having to re-approve it after each login.

turt2live commented:
> How is the user going to be prompted to accept a widget? What information will be presented to them? This implies they need to see and approve the contents of the 'data' object too, and therefore know how any given widget will behave (ie. that 'custom widget' will load the page in curl). How is the user going to make a decision on this?

This is all for the user story and out of scope here. This proposal is just to track the information, and the requirement down the pipe is that every edit needs to be tracked as a new flag (which includes `data`, cosmetics, etc).

> I can't help but think that if the user is going to have to re-approve the widget every time it's edited, this is only incrementally better than them having to re-approve it after each login.

This thought has been brought up about 20 times but doesn't seem to be a hard enough blocker.

dbkr commented:
OK - will take it to the user story

turt2live committed:
It's less verbose than we think

BillCarsonFr commented on spec/settings.md:
````diff
@@ -109,3 +109,68 @@ should be set to `false`. This should have no bearing on whether or not widgets
+```json
+{
+    "widgets": {
+        "$1543854381213sKqbg:example.org": true
````
According to the requirements, we should re-ask permission if the widget changes. Not sure how it can change, maybe an edition of the event? In this case is it enough to just store a boolean? maybe a kind content hash also?

turt2live replied:
The event ID changes when the state event changes.

BillCarsonFr reviewed (commented):
Just a comment regarding the 'if widget changes' requirement

dbkr reviewed (approved):
Talked it through and I think basically this is fine: Tom's going to update the user story to reflect how we're using the widget whitelist to infer behaviour for thing like the 'custom widget' widget (ie. what we do now). I think this makes it OK in the sense that we're sort of doing a lot more than we strictly should in order to fix this more expediently in lieu of making changes to the the integration managers / widget spec.

turt2live commented:
and with that, all the stakeholders approve so this can unblock the user story now.
